### PR TITLE
[timeseries] Refactor scaler implementation

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:*

*Description of changes:*
- Add a new parameter `target_scaler` supported by all models. If provided, a target scaler is applied to the target time series before any model-specific preprocessing.
- Upgrade to `mlforecast==0.13`
  - This required some modifications to the preprocessing logic because of breaking changes introduced in `mlforecast-0.11`
  - Instead of hacking the scaler private API with `_get_scale_per_item`, we now rely only on the public API of the scaler
- Add a property `allowed_hyperparameters` to `AbstractTimeSeriesModel` that lists the supported hyperparameters. This is necessary to silence warnings when local models receive the `target_scaler` argument. 
  - In the future, we can extend it to all models to deal with #2969

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py [hunk 1]
```diff
@@ -167,19 +165,10 @@ def _get_mlforecast_init_args(self, train_data: TimeSeriesDataFrame, model_param
...
-        else:
-            logger.warning(
-                f"Unrecognized `scaler` {scaler_name} (supported options: ['standard', 'mean_abs', None]). Scaling disabled."
-            )
-
-        if self._scaler is not None:
+        # Support "scaler" for backward compatibility
+        scaler_type = model_params.get("target_scaler", model_params.get("scaler"))
+        if scaler_type is not None:
+            self._scaler = MLForecastScaler(scaler_type=scaler_type)
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:171]: To discuss: Previously, `DirectTabular` and `RecursiveTabular` had a `"scaler"` hyperparameter that behaved very similarly to `target_scaler`. With the current change, the models "hijack" the `target_scaler` argument and implement custom scaling option (different from what's done in `AbstractTimeSeriesModel`). This is necessary to ensure that scaling is performed after differencing (inside the MLForecast object), which usually results in better accuracy.

Does this make sense?

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py [hunk 2]
```diff
@@ -472,8 +468,9 @@ def is_quantile_model(self) -> bool:
 
     def _get_model_params(self) -> dict:
         model_params = super()._get_model_params()
-        model_params.setdefault("scaler", "mean_abs")
-        model_params.setdefault("differences", [])
+        model_params.setdefault("target_scaler", "mean_abs")
+        if "differences" not in model_params or model_params["differences"] is None:
+            model_params["differences"] = []
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:473]: This change is needed so that `{"differences": None}` is interpreted as "auto differences". Currently (without the fix), `{"differences": None}` raises an error.

File: timeseries/src/autogluon/timeseries/transforms/scaler.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    """Applies min/max scaling to each time series in the dataset."""
+
+    def _compute_loc_scale(self, target_series: pd.Series) -> Tuple[pd.Series, pd.Series]:
+        stats = target_series.abs().groupby(level=ITEMID, sort=False).agg(["min", "max"])
+        scale = (stats["max"] - stats["min"]).clip(lower=self.min_scale)
+        loc = stats["min"] / scale
+        return loc, scale
+
+
+class LocalRobustScaler(LocalTargetScaler):
```
Comment (nathanaelbosch) [inline timeseries/src/autogluon/timeseries/transforms/scaler.py:86]: All scalers above have a docstring. Is there a short one-line description that could add some information here?

File: timeseries/src/autogluon/timeseries/models/gluonts/torch/models.py [hunk 4]
```diff
@@ -85,6 +85,8 @@ class DeepARModel(AbstractGluonTSModel):
         If True, ``lightning_logs`` directory will NOT be removed after the model finished training.
     """
 
+    # TODO: Should we rename "scaling: bool" to, e.g., "window_scaler": {"mean_abs", None} and have a hyperparameter names for all models?
+
```
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply timeseries/src/autogluon/timeseries/models/gluonts/torch/models.py:89]: I am not very familiar with the details, but the categorization into local scaling and window scaling seems very plausible to me.

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py (hunk 1 above)
Comment (nathanaelbosch) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:171]: I am not quite sure what "This is necessary to ensure that scaling is performed after differencing" means, but overall it seems plausible to me that we transform the previously-used scaler into an `MLForecastScaler` now.

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py [hunk 5]
```diff
@@ -553,15 +555,10 @@ def _postprocess_predictions(self, predictions: np.ndarray, repeated_item_ids: p
             predictions["mean"] = predictions["0.5"]
         else:
             predictions = pd.DataFrame(predictions, columns=["mean"])
-            predictions = self._add_gaussian_quantiles(predictions, repeated_item_ids=repeated_item_ids)
```
Comment (nathanaelbosch) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:556]: Why did this part change with the scaling?

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py (hunk 2 above)
Comment (nathanaelbosch) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:473]: What does "auto differences" mean?

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/transforms.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        self.ag_scaler = get_target_scaler(name=self.scaler_type, target=self.target_col)
+        transformed = self.ag_scaler.fit_transform(self._df_to_tsdf(df)).reset_index()
+        return self._tsdf_to_df(transformed)
+
+    def inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        transformed = self.ag_scaler.inverse_transform(self._df_to_tsdf(df))
+        return self._tsdf_to_df(transformed)
+
+
+def apply_inverse_transform(
```
Comment (nathanaelbosch) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/transforms.py:43]: I was a bit surprised at first about this function not being inside of the class above, but I suppose that this comes from the fact that we need to handle both `BaseTargetTransform` and `_BaseGroupedArrayTargetTransform`?

**Review by nathanaelbosch — COMMENTED** [review]
Thanks for the PR, and in particular also for your comments! As you know I am not too familiar with the codebase, but I left some comments and questions and hopefully resolving them will help make sure that we caught everything.

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 7]
```diff
@@ -241,11 +243,31 @@ def fit(
         model: AbstractTimeSeriesModel
             The fitted model object
         """
+        self.initialize(**kwargs)
+        self.target_scaler = self._get_target_scaler()
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py]: I see two reasons for this choice:
1. `target_scaler` is a special kind of preprocessing, since we also need to apply the `inverse_transform` on the predictions. In contrast, `preprocess` only modifies the data before it's consumed by the model.
2. `self.preprocess` contains model-specific preprocessing logic that can be modified by the subclasses. By keeping the `target_scaler` out of it, we don't need to rely on the person implementing the model to remember using the `target_scaler` appropriately.

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py (hunk 2 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:473]: "auto differences" = we set differences to `[get_seasonality(model.freq)]` by default.

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py (hunk 5 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:556]: Previously, we had a rather convoluted logic that can be summarized as

- in fit(): original data -> scaled data -> compute residuals -> save "scaled" residuals
- in predict(): original data -> scaled data -> scaled predictions -> add scaled residuals -> undo scaling

We now simplified this to
- in fit(): original data -> scaled data -> compute residuals -> undo scaling on residuals -> save residuals that are on the same scaled as original data
- in predict(): original data -> scaled data -> scaled predictions -> undo scaling -> add residuals

The latter approach is slightly cleaner and does not requires us to mess with the internals of the scaler.

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/transforms.py (hunk 6 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/transforms.py:43]: Exactly, we need a function that works both with `BaseTargetTransform` and `_BaseGroupedArrayTargetTransform` classes that have different expectations for the input format: the former works on `pd.DataFrame` while the latter only works on `GroupedArray`.

File: timeseries/src/autogluon/timeseries/transforms/scaler.py (hunk 3 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/transforms/scaler.py:92]: Good idea, added some docstrings.

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 8]
```diff
@@ -241,11 +243,31 @@ def fit(
...
+                val_data = self.target_scaler.transform(val_data)
             val_data = self.preprocess(val_data, is_train=False)
         return super().fit(train_data=train_data, val_data=val_data, **kwargs)
 
+    @property
+    def allowed_hyperparameters(self) -> List[str]:
+        """List of hyperparameters allowed by the model."""
+        return ["target_scaler"]
+
+    def _get_target_scaler(self) -> Optional[LocalTargetScaler]:
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py]: The purpose of this method is to create a `LocalTargetScaler` object based on the value of the `self.params["target_scaler"]` hyperparameter. I guess that a better name would be `_initialize_target_scaler` or `_create_target_scaler`, changing this.

There is a slightly complicated logic of how the hyperparameters are processed by the model. For example, we cannot just check the value of `self.params["target_scaler"]` during `__init__` and create the scaler there. The reason is that `self.params` might contain search spaces, and get the "concrete" hyperparameter value only after we call `self.initialize()` inside `fit()`. For this reason, we can only assign the `target_scaler` attribute during `fit()`.

[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py]: I've renamed the method to `_create_target_scaler` and added a docstring explaining the purpose.

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py]: This seems plausible to me, thanks a lot for clarifying!

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py (hunk 5 above)
Comment (nathanaelbosch) [reply timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:556]: Thank you for clarifying!

**Review by Innixma — APPROVED** [review]
LGTM!

Note that I'm not the most familiar with these changes, as it is somewhat hard to fully understand given the amount of code edits, but assuming benchmark results look fine and tests pass, should be ok.
